core/src/main/java/com/linecorp/armeria/client/endpoint/FileWatcherRegistry.java

```diff
@@ -0,0 +1,228 @@
+     * @param callback function which is invoked when file is changed.
+     *
+     * @return a key which is used to unregister from watching.
+     */
+    synchronized FileWatchRegisterKey register(Path filePath, Runnable callback) {
+        final FileWatchRegisterKey watchRegisterKey = new FileWatchRegisterKey(filePath);
+        final FileWatchServiceContext watchServiceContext = fileSystemWatchServiceMap.computeIfAbsent(
+                filePath.getFileSystem(), fileSystem -> {
+                    try {
+                        return new FileWatchServiceContext(
+                                "armeria-file-watcher-" + fileSystem.getClass().getName(),
+                                fileSystem.newWatchService());
+                    } catch (IOException e) {
+                        throw new IllegalArgumentException(
+                                "invalid filesystem for path: " + watchRegisterKey.getFilePath());
+                    }
+                });
+        watchServiceContext.register(watchRegisterKey, callback);
+        return watchRegisterKey;
+    }
+
+    /**
+     * Stops watching a properties file corresponding to the {@link FileWatchRegisterKey}. Nothing
+     * happens if the {@link FileWatchRegisterKey} is not registered or already unregistered. This
+     * method is thread safe.
+     * @param watchRegisterKey key that was used to register for watching a file.
+     */
+    synchronized void unregister(FileWatchRegisterKey watchRegisterKey) {
+        final FileSystem fileSystem = watchRegisterKey.getFilePath().getFileSystem();
+        final FileWatchServiceContext watchServiceContext = fileSystemWatchServiceMap.get(fileSystem);
+        if (watchServiceContext == null) {
+            return;
+        }
+        watchServiceContext.unregister(watchRegisterKey);
+        if (!watchServiceContext.isRunning()) {
+            fileSystemWatchServiceMap.remove(fileSystem);
+        }
+    }
+
+    /**
+     * Returns whether the current registry is watching a file.
+     * @return true if a file is watched
+     */
+    @VisibleForTesting
+    boolean isRunning() {
+        return fileSystemWatchServiceMap.values().stream().anyMatch(FileWatchServiceContext::isRunning);
+    }
+
+    /**
+     * Closes the {@link WatchService}, thread, and clear registry.
+     * @throws Exception may be thrown if an I/O error occurs
+     */
+    @Override
+    public synchronized void close() throws Exception {
+        fileSystemWatchServiceMap.values().forEach(context -> {
+            try {
+                context.close();
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
+            }
+        });
+        fileSystemWatchServiceMap.clear();
+    }
+
+    /**
+     * Key used to register/unregister for file change events.
+     */
+    static final class FileWatchRegisterKey {
+        private final Path filePath;
```

Comment:
Question: Did you wrap `Path` with `FileWatchRegisterKey` for identity comparison?

Reply:
Yes. This was done to allow duplicate path registration. Let me know if I am missing anything (different design pattern, explicitly override to show identity comparison, etc.)

Reply:
I see. How about adding some comment about identity comparison in the class Javadoc?